rmw_fastrtps_cpp/src/rmw_publisher.cpp

```diff
@@ -103,33 +103,27 @@ rmw_create_publisher(
   auto common_context = static_cast<rmw_dds_common::Context *>(node->context->impl->common);
 
   auto info = static_cast<const CustomPublisherInfo *>(publisher->data);
-  {
-    // Update graph
-    std::lock_guard<std::mutex> guard(common_context->node_update_mutex);
-    rmw_dds_common::msg::ParticipantEntitiesInfo msg =
-      common_context->graph_cache.associate_writer(
-      info->publisher_gid, common_context->gid, node->name, node->namespace_);
-    rmw_ret_t rmw_ret = rmw_fastrtps_shared_cpp::__rmw_publish(
-      eprosima_fastrtps_identifier,
-      common_context->pub,
-      static_cast<void *>(&msg),
-      nullptr);
+
+  // Update graph
+  rmw_ret_t rmw_ret = common_context->update_publisher_graph(
+    info->publisher_gid,
+    node->name, node->namespace_
+  );
+
+  if (RMW_RET_OK != rmw_ret) {
+    rmw_error_state_t error_state = *rmw_get_error_state();
+    rmw_reset_error();
+    rmw_ret = rmw_fastrtps_shared_cpp::destroy_publisher(
+      eprosima_fastrtps_identifier, participant_info, publisher);
```

Comment:
I almost feel like we should make this an `RCPPUTILS_SCOPE_EXIT` right after we successfully create the publisher above.  That way it will be destroyed in all error paths.  Thoughts?

(the same goes for the subscription and the `dynamic` versions below)